### ISSUE TITLE
npm install git repos dependencies with git submodules

### ISSUE THREAD

**Issue by stefanocudini** [npm/npm#6700]
I want use npm install for clone each git submodule(hosted on github).

party of my package.json file:
```
    "dependencies": {
        "jquery": "~2.1.0",
        "module-ui-DSDEditor": "git+https://github.com/modules-lib/module-ui-DSDEditor.git",
        "module-ui-DataEditor": "git+https://github.com/modules-lib/module-ui-DataEditor.git",
        "module-ui-analysis": "git+https://github.com/modules-lib/module-ui-analysis.git",
        "module-ui-catalog": "git+https://github.com/modules-lib/module-ui-catalog.git",
        "module-ui-metadata-editor": "git+https://github.com/modules-lib/module-ui-metadata-editor.git"
    }
```

When I run *npm install* download into *node_modules* only repo files.. without **.git** directory and GIT metadata info..

I have a **.gitmodules** files contains:
```
[submodule "node_modules/module-ui-catalog"]
	path = node_modules/module-ui-catalog
	url = https://github.com/modules-lib/module-ui-catalog.git
[submodule "node_modules/module-ui-DSDEditor"]
	path = node_modules/module-ui-DSDEditor
	url = https://github.com/modules-lib/module-ui-DSDEditor.git
[submodule "node_modules/module-ui-metadata-editor"]
	path = node_modules/module-ui-metadata-editor
	url = https://github.com/modules-lib/module-ui-metadata-editor.git
[submodule "node_modules/module-ui-DataEditor"]
	path = node_modules/module-ui-DataEditor
	url = https://github.com/modules-lib/module-ui-DataEditor.git
[submodule "node_modules/module-ui-analysis"]
	path = node_modules/module-ui-analysis
	url = https://github.com/modules-lib/fenix-ui-analysis.git
```

**Comment by smikes** [comment]
Is this still a problem for you?  It sounds like you have found a workable solution.

We are trying to clean up older npm issues, so if we don't hear back from you within a week, we will close this issue. (Don't worry -- you can always come back again and open a new issue!)

Thanks!

**Comment by stefanocudini** [comment]
no io non ho trovato nessun altra soluzione per ora, quando la troverò
riaprirò la issue
No at this time I not found one best solution.. when I will find then I
will reopen this issue

thanks
Stefano

2015-01-21 13:43 GMT+01:00 Sam Mikes <notifications@github.com>:

**Comment by othiym23** [comment]
Git repos are never cloned directly into your application. They always follow this process:

1. npm does a bare clone of the repo into `$TMPDIR`
2. npm clones that clone into a directory in your npm cache, and then checks out the treeish that the branch requested mapped to.
3. npm produces an installable package tarball from that checkout that is then placed into your npm cache.
4. that cached tarball is installed into your application.

If you want to be working directly from the Git checkout of your project, with all its submodules intact, your best bet is probably to `npm link` for the project with the submodules, and then `npm link <project name>` it into the package consuming it. It's more unwieldy than using Git URLs in your `package.json`, but it will give you the control you need about how files are laid out.

**Comment by motin** [comment]
@othiym23 I quoted your anser on http://stackoverflow.com/questions/23527393/can-npm-install-the-git-folder-for-dependencies/28559920#28559920 hope it's ok. 
Such a shame that npm won't handle this.

**Comment by othiym23** [comment]
@motin I feel like the needs of package management and version control are usefully separable here. `npm link` gives you the power to work on a live version of your package while still having it in source control while remaining agnostic to the underlying version control system, and that seems like the appropriate solution to the problem. The other solutions people have come up with on this issue are a little cumbersome, but that's because, in my opinion, git submodules themselves are somewhat unwieldy. I use them myself, in a few projects, but only in cases (like my vim configuration) where there wasn't a really good alternative.

npm makes calls to git it when it's convenient and helpful, but npm is not meant to be a layer on top of (or a replacement for) Git.

Closing as resolved, or as resolved as it's going to get.

**Comment by Joshuabaker2** [comment]
For posterity, I found that my current version of npm (3.8.9) now does install submodules. It appears that the feature was added: 
https://github.com/npm/npm/commit/39dea9ca4216c6ea628f5ca47d2b34a4b251a1ed

**Comment by RacingTadpole** [comment]
I know this is an old and closed issue, but in case it helps anyone else who lands here, another solution is [npmgitdev](https://www.npmjs.com/package/npmgitdev). This is a wrapper which lets you work directly with git repos inside `node_modules`.

**Comment by gdamjan** [comment]
@Joshuabaker2 
in what scenario does it work for you? what's your setup

**Comment by Joshuabaker2** [comment]
This was a while ago, so I don't remember exactly - but I believe what I had set up was a git repository as an npm dependency in my package.json. That repo had git submodules, and with older versions of npm those submodules needed to be manually `init`'d and `update`d. However, with the more recent versions of npm (such as 3.8.9), those submodules were automatically `init`'d and `update`d.

**Comment by fjcaetano** [comment]
I know this is an old and closed issue, but npm version 4.2.0 is not installing submodules for dependencies registered pointing to a git repository.

If I register a dependency pointing to a version it installs the submodules as expected.

Is that the intended behavior?

**Comment by ganqqwerty** [comment]
I have this issue with npm 5.10.0 when in package.json I added     `"ng-swagger-gen": "github:ganqqwerty/ng-swagger-gen#1a53f13"` instead of `"ng-swagger-gen": "ganqqwerty/ng-swagger-gen#1a53f13"`. Weird
